Add Sphinx to build documentation for DLPack

tirthasheshpatel opened the pull request:
Hi everyone,

I am Tirth, an intern at [Quansight Labs](https://www.quansight.com/labs). I will be working on DLPack under the mentorship of @mattip to add documentation, tests, and contributing bug-fixes/features to other packages that use DLPack (e.g. NumPy, CuPy) and DLPack itself.

This PR introduces [Sphinx](https://www.sphinx-doc.org/en/master/) as a documentation tool for DLPack. It uses the [Breathe](https://breathe.readthedocs.io/en/latest/) extension to parse Doxygen output (because Sphinx doesn't directly parse C/C++). The new documentation uses the [PyData Sphinx Theme](https://pydata-sphinx-theme.readthedocs.io/en/latest/) (adopted by several projects in the data science community). To keep the diff minimal, I didn't change the default configuration but it can be changed later to add sidebars, logos, social media links, etc. This PR also adds a small CI script to build docs (just to test if Make and CMake changes work properly).

I have deployed the documentation [here](https://tirthasheshpatel.github.io/dlpack/) for testing. In a follow-up PR, we can add a script to deploy the docs on `gh-pages` branch of the main repo. Then we can start populating the docs with examples, tutorials, list of projects that use DLPack, etc.

I would appreciate your feedback and thought on this PR, thanks!

tirthasheshpatel committed:
Add Sphinx to build documentation for DLPack

This PR introduces Sphinx as a documentation tool for DLPack. It uses the
Breathe extension to parse Doxygen output (because Sphinx doesn't directly
parse C/C++). The new documentation uses the PyData sphinx theme (adopted
by several projects in the data science community). To keep the diff minimal,
I didn't change the default configuration but it can be changed later to add
sidebars, logo, social media links, etc.

I have deployed the documentation [here](https://tirthasheshpatel.github.io/dlpack/)
for testing. In a follow-up PR, we can add a script that deploys the docs on
`gh-pages` branch of the main repo. Then we can start populating the docs with
examples, tutorials, list of projects that use DLPack, etc.

This PR also adds a small CI script to build docs (just to test Make and CMake
changes work properly).

mattip commented on doc_requirements.txt:
```diff
@@ -0,0 +1,3 @@
+Sphinx
+git+https://github.com/pydata/pydata-sphinx-theme.git@master
+breathe
```
I prefer to pin versions

```suggestion
Sphinx==4.4.0
pydata-sphinx-theme==.0.8.0
breathe==4.31.0
```

tirthasheshpatel committed:
Pin versions

tirthasheshpatel committed:
Use a previous stable version of pydata sphinx theme

tirthasheshpatel committed:
Address code reviews

- Change `show_docs` target to use http.server instead of webbrowser
- Comment out `html_static_path` and `html_templates_path` to avoid Sphinx Warnings
- Use the default Python 3.8 in CI

tirthasheshpatel committed:
Add instructions in docs/README.md to build documentation

tirthasheshpatel committed:
Fix a typo in the docs/README.md file

tirthasheshpatel committed:
doc_requirement.txt -> doc_requirements.txt in docs/README.md
